src/Amalgam/string/StringInternPool.h

```diff
@@ -194,96 +194,51 @@ class StringInternPool
 		ValidateStringIdExistence(id);
 	#endif
 
-		int64_t refcount = id->refCount--;
-
-		//if other references, then can't clear it; signed, so it won't wrap around
-		if(refcount > 1)
-			return;
-
 	#if defined(MULTITHREAD_SUPPORT) || defined(MULTITHREAD_INTERFACE)
-		//this thread is about to free the reference, but need to acquire a lock
-		// so, keep the reference alive by incrementing it *before* attempting the lock
-		id->refCount++;
+		//decrement the counter in an atomic fashion
+		//but if down to the last reference, then need to acquire a lock
+		while(true)
+		{
+			int64_t ref_count = id->refCount.load();
+			if(ref_count <= 1)
+				break;
+
+			//if can decrement, return
+			if(std::atomic_compare_exchange_weak(&id->refCount, &ref_count, ref_count - 1))
+				return;
+		}
 
-		Concurrency::SingleLock lock(mutex);
+		Concurrency::Lock lock(mutex);
 
-		//with the lock, decrement reference count in case this string should stay active
-		refcount = id->refCount--;
+		//retry decrementing in case it has been modified by another thread
+		while(true)
+		{
+			int64_t ref_count = id->refCount.load();
+			if(ref_count <= 1)
+				break;
 
-		//if other references, then can't clear it
-		if(refcount > 1)
+			//if can decrement, return
+			if(std::atomic_compare_exchange_weak(&id->refCount, &ref_count, ref_count - 1))
+				return;
+		}
+	#else
+		//remove any that aren't the last reference
+		id->refCount--;
+		if(id->refCount > 0)
 			return;
 	#endif
-
+		
 		stringToID.erase(id->string);
 	}
 
-
 	//creates new references from the references container and function
 	template<typename ReferencesContainer,
 		typename GetStringIdFunction = StringID(StringID)>
 	inline void DestroyStringReferences(ReferencesContainer &references_container,
 		GetStringIdFunction get_string_id = [](auto sid) { return sid;  })
 	{
-	#if !defined(MULTITHREAD_SUPPORT) && !defined(MULTITHREAD_INTERFACE)
 		for(auto r : references_container)
 			DestroyStringReference(get_string_id(r));
-	#else
-		if(references_container.size() == 0)
-			return;
-
-		//as it goes through, if any id needs removal, will set this to true so that
-		// removal can be done after reference count decreases are done
-		bool ids_need_removal = false;
-
-		for(auto r : references_container)
-		{
-			StringID id = get_string_id(r);
-			if(id == NOT_A_STRING_ID || id == emptyStringId)
-				continue;
-
-		#ifdef STRING_INTERN_POOL_VALIDATION
-			ValidateStringIdExistence(id);
-		#endif
-
-			int64_t refcount = id->refCount--;
-
-			//if extra references, just return, but if it is 1, then it will try to clear
-			if(refcount <= 1)
-				ids_need_removal = true;
-		}
-
-		if(!ids_need_removal)
-			return;
-
-		//need to remove at least one reference, so put all counts back while wait for lock
-		for(auto r : references_container)
-		{
-			StringID id = get_string_id(r);
-			if(id != NOT_A_STRING_ID && id != emptyStringId)
-				id->refCount++;
-		}
-
-		Concurrency::SingleLock lock(mutex);
```

Comment:
Noting that the previous version of this gets the lock only once, regardless of how many IDs are deleted; this change gets the lock once per ID.  I don't know if that's a significant performance cost.